src/inmanta/server/bootloader.py

```diff
@@ -82,7 +83,16 @@ async def start(self) -> None:
         await self.restserver.start()
         self.started = True
 
-    async def stop(self) -> None:
+    async def stop(self, timeout: Optional[int] = None) -> None:
+        """
+        :param timeout: Raises TimeoutError when the server hasn't finished stopping after
+                        this amount of seconds.
+        """
+        if not timeout:
+            await self._stop()
+        await asyncio.wait_for(self._stop(), timeout=timeout)
+
+    async def _stop(self) -> None:
```

Comment:
Could you elaborate on the `cancel` part? What is your concern exactly?

Reply:
[`wait_for` ](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for)

`If a timeout occurs, it cancels the task and raises [asyncio.TimeoutError](https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.TimeoutError).`

So the wait_for can cause the stop process itself to be canceled.

Reply:
True. But if we don't do the shield, we get a stack trace that references the task that was hanging, because it fails with a `CancelledError`. If we do the shield, the stack trace only goes `asyncio.wait_for()` statement. That way we don't have any debugging information. We could also use `asyncio.all_tasks()` to obtain that information, but I am not sure whether that's any better. What do you think?

Reply:
With the warning, it is fine for me.